aten/src/ATen/native/mkldnn/xpu/detail/WoQMatmul.cpp

```diff
@@ -0,0 +1,178 @@
+#include <c10/xpu/XPUFunctions.h>
+
+#include <ATen/native/mkldnn/xpu/detail/Attr.h>
+#include <ATen/native/mkldnn/xpu/detail/Utils.h>
+
+#include <oneapi/dnnl/dnnl.hpp>
+#include <cstdint>
+
+namespace at::native::onednn {
+
+sycl::event woq_matmul_int4(
+    Tensor& result, // torchao: [M, K], dtype: fp16
+    const Tensor& mat1_, // torchao: [M, K], dtype: fp16
+    const Tensor& mat2_, // torchao quantized weight, [K/8, N], dtype: uint4x8
+    const Tensor& scale, // torchao: [K/group_size, N], dtype: fp16
+    const Tensor& zp, // torchao: [K/group_size, N], dtype: int8
+    int64_t group_size,
+    Attr attr,
+    const std::vector<sycl::event>& deps) {
+  size_t dims = result.dim();
+  TORCH_CHECK(
+      dims == 2, "INT4 matmul at XPU only works with 2D input, got ", dims);
+  TORCH_CHECK(result.defined(), "oneDNN matmul result should be defined");
+
+  at::Device cur_device = at::Device(at::kXPU, at::xpu::current_device());
+  auto engine = GpuEngineManager::Instance().get_engine(cur_device);
+  auto stream = GpuStreamManager::Instance().get_stream();
+
+  // make them all contiguous
+  Tensor m1 = mat1_;
+  Tensor m2 = mat2_;
+  Tensor scale_ = scale;
+  Tensor zp_ = zp;
+  Tensor dst = result;
+
+  int m = m1.size(-2); // M
+  int n = dst.size(-1); // N
+  int k = m1.size(-1); // K
+
+  // Construct usr md from input
+  // xxx_usr_md would describe the real layout of inputs
+  auto m1_usr_dt = get_onednn_dtype(m1); // e.g., half <==> f16
+  auto m2_usr_dt = get_onednn_dtype(m2); // int32 tensor, pack 8 int4
+  auto scale_usr_dt = get_onednn_dtype(scale_); // bf16
+  auto zp_usr_dt = get_onednn_dtype(zp_); // s8 expected currently
+  auto dst_usr_dt = get_onednn_dtype(dst); // bf16
+
+  dnnl::memory::dims m1_usr_dims, m2_usr_dims, scale_usr_dims, zp_usr_dims,
+      dst_usr_dims;
+  dnnl::memory::dims m1_usr_strides, m2_usr_strides, scale_usr_strides,
+      zp_usr_strides, dst_usr_strides;
+  int64_t compressed_k = (uint64_t)(k / 8);
```

Comment:
why you cast `(k/8)` to `uint64_t` but define `compressed_k` as `int64_t`.

Reply:
silly of me : ), I align the dtype here now.